R/logistic_reg.R

```diff
@@ -0,0 +1,342 @@
+    )
+  }
+  if (any(sample_weight < 0) || any(!is.finite(sample_weight))) {
+    stop(
+      "Expected `sample_weight` to only contain finite, non-negative, and ",
+      "non-NaN numeric values."
+    )
+  }
+}
+
+logistic_reg_build_sample_weight <- function(sample_weight,
+                                             class_weight,
+                                             processed) {
+  outcomes <- processed$outcomes[[1]]
+  n_samples <- length(outcomes)
+  if (identical(class_weight, "balanced")) {
+    class_weight <- n_samples / table(outcomes)
+  }
+  if (!is.null(class_weight)) {
+    # update each sample weight value by multiplying it with its corresponding
+    # class weight
+    sample_weight <- sample_weight *
+      as.numeric(class_weight[levels(outcomes)[outcomes]])
+  }
+
+  sample_weight
+}
+
+#' Train a logistic regression model.
+#'
+#' Train a logistic regression model using Quasi-Newton (QN) algorithms (i.e.,
+#' Orthant-Wise Limited Memory Quasi-Newton (OWL-QN) if there is L1
+#' regularization, Limited Memory BFGS (L-BFGS) otherwise).
+#'
+#' @template model-with-numeric-input
+#' @template fit-intercept
+#' @param penalty The penalty type, must be one of
+#'   {"none", "l1", "l2", "elasticnet"}.
+#'   If "none" or "l2" is selected, then L-BFGS solver will be used.
+#'   If "l1" is selected, solver OWL-QN will be used.
+#'   If "elasticnet" is selected, OWL-QN will be used if l1_ratio > 0, otherwise
+#'   L-BFGS will be used. Default: "l2".
+#' @param tol Tolerance for stopping criteria. Default: 1e-4.
+#' @param C Inverse of regularization strength; must be a positive float.
+#'   Default: 1.0.
+#' @param class_weight If \code{NULL}, then each class has equal weight of
+#'   \code{1}.
+#'   If \code{class_weight} is set to \code{"balanced"}, then weights will be
+#'   inversely proportional to class frequencies in the input data.
+#'   If otherwise, then \code{class_weight} must be a named numeric vector of
+#'   weight values, with names being class labels.
+#'   If \code{class_weight} is not \code{NULL}, then each entry in
+#'   \code{sample_weight} will be adjusted by multiplying its original value
+#'   with the class weight of the corresponding sample's class.
+#'   Default: NULL.
+#' @param sample_weight Array of weights assigned to individual samples.
+#'   If \code{NULL}, then each sample has an equal weight of 1. Default: NULL.
+#' @param max_iters Maximum number of solver iterations. Default: 1000L.
+#' @param linesearch_max_iters Max number of linesearch iterations per outer
+#'   iteration used in the LBFGS- and OWL- QN solvers. Default: 50L.
+#' @param l1_ratio The Elastic-Net mixing parameter, must \code{NULL} or be
+#'   within the range of [0, 1]. Default: NULL.
+#'
+#' @examples
+#' library(cuda.ml)
+#'
+#' X <- scale(as.matrix(iris[names(iris) != "Species"]))
+#' y <- iris$Species
+#'
+#' model <- cuda_ml_logistic_reg(X, y, max_iters = 100)
```

Comment:
Also, can you please expand the example to show how to get the equivalent result with `glm()`?